spec/Candid.md

```diff
@@ -998,10 +1013,19 @@ C[variant { <nat> = <t>; _;* } <: variant { <nat> = <t'>; _;* }](variant { <nat>
 
 #### Dynamic
 
-On dynamic types, coercion is again the identity:
+On the dynamic type, coercion is the identity:
 ```
 C[dynamic <: dynamic](x) = x
 ```
+Any data type can be coerced to `dynamic`:
+```
+C[<t> <: dynamic](<v>) = dynamic <v> : <t>  if <t> =/= dynamic
+```
+The inverse direction is only possible if the encapsulated value matches the target type, such that the corresponding coercions is defined:
+```
+C[dynamic <: <t>](dynamic <v'> : <t'>) = C[<t'> <: <t>](<v'>)  if <t> =/= dynamic
+```
+Note: Type `<t>` is not known statically, so it cannot be decided statically whether `C[<t'> <: <t>]` is defined. Hence this amouts to a runtime type check.
```

Comment:
That's a bit too sneaky; if we go this route we should explicitly define this function to be partial (it was total before), with suitable notation. Also the interaction with `opt` stuff needs to be clear, especially if we go back to making _that_ more dynamic again.

How does this affect our formal guarantees? They can only hold when no `dynamic` is in use, right?

Reply:
Fair enough. It's probably necessary to distinguish gradual conversion from subtyping anyway, because transitively, subtyping collapses with the above rules, i.e., everything becomes a subtype of everything.

Reply:
@nomeata, you are right, this was taken at least one shortcut too many. I pushed a change that should properly separate graduality from subtyping and coercions. Subtyping now only allows `t <: dynamic`, but not the other way round. Graduality itself is handled in de/serialisation itself. Let me know what you think.